{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "try:\n",
    "    import itertools.izip as zip\n",
    "except ImportError:\n",
    "    import itertools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_size = 1000\n",
    "alphabet = ['0', '1']\n",
    "seq_length = 1\n",
    "\n",
    "base_haplotype = \"0\"\n",
    "fraction_base = 0.5\n",
    "\n",
    "fitness_nodrug = {'base': 1, 'resistant': 0.8, 'fusion': 0.9}\n",
    "fitness_drug = {'base': 0.8, 'resistant': 1.0, 'fusion': 0.9}\n",
    "\n",
    "fitness_base = fitness_nodrug['base']\n",
    "fitness_resistant = fitness_nodrug['resistant']\n",
    "fitness_fusion = fitness_nodrug['fusion']\n",
    "\n",
    "fitness_effect = 1 # fitness effect if a functional mutation occurs\n",
    "fitness_chance = 0.01 # chance that a mutation has a fitness effect\n",
    "\n",
    "mutation_rate = 0.005\n",
    "fusion_rate = 0.01\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting population: {'0': 500.0, '1': 500.0}\n",
      "Fitness: {'0': 1, '1': 0.8}\n"
     ]
    }
   ],
   "source": [
    "pop = {}\n",
    "pop[\"0\"] = pop_size * fraction_base\n",
    "pop[\"1\"] = pop_size * (1 - fraction_base)\n",
    "\n",
    "print(\"Starting population: {}\".format(pop))\n",
    "\n",
    "fitness = {}\n",
    "fitness[\"0\"] = fitness_base\n",
    "fitness[\"1\"] = fitness_resistant\n",
    "\n",
    "print(\"Fitness: {}\".format(fitness))\n",
    "\n",
    "def get_random_haplotype():\n",
    "    \"\"\"Returns a random haplotype from the population, weighted by the frequency of the haplotype\"\"\"\n",
    "    haplotypes = list(pop.keys()) \n",
    "    frequencies = [x/float(pop_size) for x in pop.values()]\n",
    "    total = sum(frequencies)\n",
    "    frequencies = [x / total for x in frequencies]\n",
    "    return np.random.choice(haplotypes, p=frequencies)\n",
    "\n",
    "def update_fitness(haplotype):\n",
    "    old_fitness = fitness[haplotype]\n",
    "    if (np.random.random() < fitness_chance):\n",
    "        return old_fitness * fitness_effect\n",
    "    else:\n",
    "        return old_fitness\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mutation_count():\n",
    "    mean = mutation_rate * pop_size * seq_length\n",
    "    return np.random.poisson(mean)\n",
    "\n",
    "def make_mutant(haplotype):\n",
    "    site = np.random.randint(seq_length)\n",
    "    possible_mutations = list(alphabet)\n",
    "    possible_mutations.remove(haplotype[site])\n",
    "    mutation = np.random.choice(possible_mutations)\n",
    "    new_haplotype = haplotype[:site] + mutation + haplotype[site+1:]\n",
    "    return new_haplotype    \n",
    "\n",
    "def get_mutant_fitness(haplotype):\n",
    "    old_fitness = fitness[haplotype]\n",
    "    if (np.random.random() < fitness_chance):\n",
    "        return old_fitness * fitness_effect\n",
    "    else:\n",
    "        return old_fitness\n",
    "\n",
    "def mutation_event():\n",
    "    haplotype = get_random_haplotype()\n",
    "    \n",
    "    # Only let non-fused cells mutate.\n",
    "    if len(haplotype) > 1:\n",
    "        return\n",
    "\n",
    "    if pop[haplotype] > 1:\n",
    "        pop[haplotype] -= 1\n",
    "        new_haplotype = make_mutant(haplotype)\n",
    "        if new_haplotype in pop:\n",
    "            pop[new_haplotype] += 1\n",
    "        else:\n",
    "            pop[new_haplotype] = 1\n",
    "        if new_haplotype not in fitness:\n",
    "            fitness[new_haplotype] = get_mutant_fitness(haplotype)\n",
    "\n",
    "def mutation_step():\n",
    "    mutation_count = get_mutation_count()\n",
    "    for i in range(mutation_count):\n",
    "        mutation_event()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fusion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fusion_count():\n",
    "    mean = fusion_rate * pop_size\n",
    "    return np.random.poisson(mean)\n",
    "\n",
    "def make_fusion(hap1, hap2):\n",
    "    fused = hap1 + hap2\n",
    "    return ''.join(sorted(fused))\n",
    "\n",
    "def get_fusion_fitness(haplotype):\n",
    "    from math import e\n",
    "    \n",
    "    length = len(haplotype)\n",
    "\n",
    "    # Don't let the haplotype go past 2 chars.\n",
    "    if length > 2:\n",
    "        return 0\n",
    "\n",
    "    # Use a Gaussian function such that the fitness is lower when very short and very long.\n",
    "    # f(x) = ae^(-(x-b)^2/2c^2)\n",
    "    a = fitness_fusion\n",
    "    b = 2 # max fitness at length 2\n",
    "    c = 1 # stdev\n",
    "    \n",
    "    fitness = a * e ** (-1 * ((length - b) ** 2) / ((2 * c) ** 2))\n",
    "    return fitness\n",
    "\n",
    "def fusion_event():\n",
    "    hap1 = get_random_haplotype()\n",
    "    hap2 = get_random_haplotype()\n",
    "\n",
    "    # Don't let the same haplotypes fuse.\n",
    "    if hap1 == hap2:\n",
    "        return\n",
    "        \n",
    "    # If fusion would remove either population of cells, don't let it happen.\n",
    "    if pop[hap1] == 1 or pop[hap2] == 1:\n",
    "        return\n",
    "\n",
    "    pop[hap1] -= 1\n",
    "    pop[hap2] -= 1\n",
    "\n",
    "    new_haplotype = make_fusion(hap1, hap2)\n",
    "\n",
    "    if new_haplotype in pop:\n",
    "        pop[new_haplotype] += 1\n",
    "    else:\n",
    "        pop[new_haplotype] = 1\n",
    "    if new_haplotype not in fitness:\n",
    "        fitness[new_haplotype] = get_fusion_fitness(new_haplotype)\n",
    "\n",
    "def fusion_step():\n",
    "    fusion_count = get_fusion_count()\n",
    "    for i in range(fusion_count):\n",
    "        fusion_event()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Offspring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_offspring_counts():\n",
    "    haplotypes = list(pop.keys())\n",
    "    frequencies = [pop[haplotype]/float(pop_size) for haplotype in haplotypes]\n",
    "    fitnesses = [fitness[haplotype] for haplotype in haplotypes]\n",
    "    weights = [x * y for x,y in zip(frequencies, fitnesses)]\n",
    "    total = sum(weights)\n",
    "    weights = [x / total for x in weights]\n",
    "    return list(np.random.multinomial(pop_size, weights))\n",
    "\n",
    "def offspring_step():\n",
    "    haplotypes = list(pop.keys())\n",
    "    counts = get_offspring_counts()\n",
    "    for (haplotype, count) in zip(haplotypes, counts):\n",
    "        if (count > 0):\n",
    "            pop[haplotype] = count\n",
    "        else:\n",
    "            del pop[haplotype]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Levenshtein import distance as levenshtein_distance\n",
    "\n",
    "def get_distance(seq_a, seq_b):\n",
    "    return levenshtein_distance(seq_a, seq_b)\n",
    "\n",
    "def get_diversity(population):\n",
    "    haplotypes = list(population.keys())\n",
    "    haplotype_count = len(haplotypes)\n",
    "    diversity = 0\n",
    "    for i in range(haplotype_count):\n",
    "        for j in range(haplotype_count):\n",
    "            haplotype_a = haplotypes[i]\n",
    "            haplotype_b = haplotypes[j]\n",
    "            frequency_a = population[haplotype_a] / float(pop_size)\n",
    "            frequency_b = population[haplotype_b] / float(pop_size)\n",
    "            frequency_pair = frequency_a * frequency_b\n",
    "            diversity += frequency_pair * get_distance(haplotype_a, haplotype_b)\n",
    "    return diversity\n",
    "\n",
    "def get_diversity_trajectory():\n",
    "    trajectory = [get_diversity(generation) for generation in history]\n",
    "    return trajectory\n",
    "\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "\n",
    "def diversity_plot():\n",
    "    mpl.rcParams['font.size']=14\n",
    "    trajectory = get_diversity_trajectory()\n",
    "    plt.plot(trajectory, \"#447CCD\")    \n",
    "    plt.ylabel(\"diversity\")\n",
    "    plt.xlabel(\"generation\")\n",
    "\n",
    "def get_divergence(population):\n",
    "    haplotypes = population.keys()\n",
    "    divergence = 0\n",
    "    for haplotype in haplotypes:\n",
    "        frequency = population[haplotype] / float(pop_size)\n",
    "        divergence += frequency * get_distance(base_haplotype, haplotype)\n",
    "    return divergence\n",
    "\n",
    "def get_divergence_trajectory():\n",
    "    trajectory = [get_divergence(generation) for generation in history]\n",
    "    return trajectory\n",
    "\n",
    "def divergence_plot():\n",
    "    mpl.rcParams['font.size']=14\n",
    "    trajectory = get_divergence_trajectory()\n",
    "    plt.plot(trajectory, \"#447CCD\")\n",
    "    plt.ylabel(\"divergence\")\n",
    "    plt.xlabel(\"generation\") \n",
    "\n",
    "def get_frequency(haplotype, generation):\n",
    "    pop_at_generation = history[generation]\n",
    "    if haplotype in pop_at_generation:\n",
    "        return pop_at_generation[haplotype]/float(pop_size)\n",
    "    else:\n",
    "        return 0\n",
    "    \n",
    "def get_trajectory(haplotype):\n",
    "    trajectory = [get_frequency(haplotype, gen) for gen in range(generations)]\n",
    "    return trajectory\n",
    "\n",
    "def get_all_haplotypes():\n",
    "    haplotypes = set()   \n",
    "    for generation in history:\n",
    "        for haplotype in generation:\n",
    "            haplotypes.add(haplotype)\n",
    "    return haplotypes\n",
    "\n",
    "colors = [\"#781C86\", \"#571EA2\", \"#462EB9\", \"#3F47C9\", \"#3F63CF\", \"#447CCD\", \"#4C90C0\", \"#56A0AE\", \"#63AC9A\", \"#72B485\", \"#83BA70\", \"#96BD60\", \"#AABD52\", \"#BDBB48\", \"#CEB541\", \"#DCAB3C\", \"#E49938\", \"#E68133\", \"#E4632E\", \"#DF4327\", \"#DB2122\"]\n",
    "colors_lighter = [\"#A567AF\", \"#8F69C1\", \"#8474D1\", \"#7F85DB\", \"#7F97DF\", \"#82A8DD\", \"#88B5D5\", \"#8FC0C9\", \"#97C8BC\", \"#A1CDAD\", \"#ACD1A0\", \"#B9D395\", \"#C6D38C\", \"#D3D285\", \"#DECE81\", \"#E8C77D\", \"#EDBB7A\", \"#EEAB77\", \"#ED9773\", \"#EA816F\", \"#E76B6B\"]\n",
    "\n",
    "def stacked_trajectory_plot(xlabel=\"generation\"):\n",
    "    mpl.rcParams['font.size']=18\n",
    "    haplotypes = get_all_haplotypes()\n",
    "    trajectories = [get_trajectory(haplotype) for haplotype in haplotypes]\n",
    "    plt.stackplot(range(generations), trajectories, colors=colors_lighter, labels=haplotypes)\n",
    "    # plt.legend(bbox_to_anchor=(1.2, 1))\n",
    "    plt.ylim(0, 1)\n",
    "    plt.ylabel(\"frequency\")\n",
    "    plt.xlabel(xlabel)\n",
    "\n",
    "def get_snp_frequency(site, generation):\n",
    "    minor_allele_frequency = 0.0\n",
    "    pop_at_generation = history[generation]\n",
    "    for haplotype in pop_at_generation.keys():\n",
    "        allele = haplotype[site]\n",
    "        frequency = pop_at_generation[haplotype] / float(pop_size)\n",
    "        if allele != \"0\":\n",
    "            minor_allele_frequency += frequency\n",
    "    return minor_allele_frequency\n",
    "\n",
    "def get_snp_trajectory(site):\n",
    "    trajectory = [get_snp_frequency(site, gen) for gen in range(generations)]\n",
    "    return trajectory\n",
    "\n",
    "def get_all_snps():\n",
    "    snps = set()   \n",
    "    for generation in history:\n",
    "        for haplotype in generation:\n",
    "            for site in range(seq_length):\n",
    "                if haplotype[site] != \"0\":\n",
    "                    snps.add(site)\n",
    "    return snps\n",
    "\n",
    "def snp_trajectory_plot(xlabel=\"generation\"):\n",
    "    mpl.rcParams['font.size']=18\n",
    "    snps = get_all_snps()\n",
    "    trajectories = [get_snp_trajectory(snp) for snp in snps]\n",
    "    data = []\n",
    "    for trajectory, color in zip(trajectories, itertools.cycle(colors)):\n",
    "        data.append(range(generations))\n",
    "        data.append(trajectory)    \n",
    "        data.append(color)\n",
    "    fig = plt.plot(*data)   \n",
    "    plt.ylim(0, 1)\n",
    "    plt.ylabel(\"frequency\")\n",
    "    plt.xlabel(xlabel)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_drug(is_present):\n",
    "    # TODO: Maybe don't use globals here? We can pass the drug state into the step functions.\n",
    "    global fitness_base\n",
    "    global fitness_resistant\n",
    "    global fitness_fusion\n",
    "\n",
    "    if is_present:\n",
    "        fitness_base = fitness_drug['base']\n",
    "        fitness_resistant = fitness_drug['resistant']\n",
    "        fitness_fusion = fitness_drug['fusion']\n",
    "    else:\n",
    "        fitness_base = fitness_nodrug['base']\n",
    "        fitness_resistant = fitness_nodrug['resistant']\n",
    "        fitness_fusion = fitness_nodrug['fusion']\n",
    "\n",
    "    fitness[\"0\"] = fitness_base\n",
    "    fitness[\"1\"] = fitness_resistant\n",
    "    haplotypes = list(pop.keys())\n",
    "    for hap in haplotypes:\n",
    "        if len(hap) < 2: continue\n",
    "        fitness[hap] = get_fusion_fitness(hap)\n",
    "\n",
    "\n",
    "# Define functions for all the regimens for convenience\n",
    "\n",
    "drug_state = False\n",
    "\n",
    "def drug_off(generation):\n",
    "    set_drug(False)\n",
    "\n",
    "def drug_on(generation):\n",
    "    set_drug(True)\n",
    "\n",
    "def drug_on_off(generation):\n",
    "    if generation == 0:\n",
    "        set_drug(True)\n",
    "    elif generation == 250:\n",
    "        set_drug(False)\n",
    "\n",
    "\n",
    "def drug_off_on(generation):\n",
    "    if generation == 0:\n",
    "        set_drug(False)\n",
    "    elif generation == 250:\n",
    "        set_drug(True)\n",
    "\n",
    "def drug_off_on_off_on(generation):\n",
    "    global drug_state\n",
    "    if generation == 0:\n",
    "        drug_state = False\n",
    "    elif generation % 150 == 0:\n",
    "        drug_state = not drug_state\n",
    "        set_drug(drug_state)\n",
    "    set_drug(drug_state)\n",
    "\n",
    "regimens = {\n",
    "    0: drug_off,\n",
    "    1: drug_on,\n",
    "    2: drug_off_on,\n",
    "    3: drug_on_off,\n",
    "    4: drug_off_on_off_on\n",
    "}\n",
    "\n",
    "# Make it easy to save the regimen name as part of the output file.\n",
    "# TODO: Maybe just make these the keys for the regimen dict?\n",
    "regimen_names = {\n",
    "    0: 'drug_off',\n",
    "    1: 'drug_on',\n",
    "    2: 'drug_off_on',\n",
    "    3: 'drug_on_off',\n",
    "    4: 'drug_off_on_off_on'\n",
    "}\n",
    "\n",
    "def drug_step(generation, regimen_num):\n",
    "    regimens[regimen_num](generation)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final population: {'0': 978, '1': 17, '01': 5}\n",
      "Final population: {'0': 979, '1': 20, '01': 1}\n",
      "Final population: {'0': 985, '1': 15}\n",
      "Final population: {'0': 982, '1': 16, '01': 2}\n",
      "Final population: {'0': 987, '1': 12, '01': 1}\n",
      "Final population: {'0': 989, '1': 10, '01': 1}\n",
      "Final population: {'0': 988, '1': 11, '01': 1}\n",
      "Final population: {'0': 983, '1': 16, '01': 1}\n",
      "Final population: {'0': 996, '1': 4}\n",
      "Final population: {'0': 985, '1': 13, '01': 2}\n",
      "Mean 01 proportion after 10 trials: 0.0014000000000000004\n",
      "Final population: {'0': 988, '1': 11, '01': 1}\n",
      "Final population: {'0': 985, '1': 13, '01': 2}\n",
      "Final population: {'0': 976, '1': 20, '01': 4}\n",
      "Final population: {'0': 991, '1': 9}\n",
      "Final population: {'0': 981, '1': 6, '01': 13}\n",
      "Final population: {'0': 974, '1': 22, '01': 4}\n",
      "Final population: {'0': 977, '1': 19, '01': 4}\n",
      "Final population: {'0': 984, '1': 16}\n",
      "Final population: {'0': 968, '1': 19, '01': 13}\n",
      "Final population: {'0': 979, '1': 18, '01': 3}\n",
      "Mean 01 proportion after 10 trials: 0.0044\n",
      "Final population: {'0': 982, '1': 15, '01': 3}\n",
      "Final population: {'0': 972, '1': 23, '01': 5}\n",
      "Final population: {'0': 971, '1': 19, '01': 10}\n",
      "Final population: {'0': 968, '1': 21, '01': 11}\n",
      "Final population: {'0': 970, '1': 18, '01': 12}\n",
      "Final population: {'0': 981, '1': 15, '01': 4}\n",
      "Final population: {'0': 968, '1': 15, '01': 17}\n",
      "Final population: {'0': 986, '1': 13, '01': 1}\n",
      "Final population: {'0': 968, '1': 18, '01': 14}\n",
      "Final population: {'0': 972, '1': 27, '01': 1}\n",
      "Mean 01 proportion after 10 trials: 0.0078\n",
      "Final population: {'0': 9, '1': 991}\n",
      "Final population: {'0': 15, '1': 984, '01': 1}\n",
      "Final population: {'0': 10, '1': 986, '01': 4}\n",
      "Final population: {'0': 22, '1': 968, '01': 10}\n",
      "Final population: {'0': 20, '1': 979, '01': 1}\n",
      "Final population: {'0': 6, '1': 994}\n",
      "Final population: {'0': 21, '1': 978, '01': 1}\n",
      "Final population: {'0': 24, '1': 973, '01': 3}\n",
      "Final population: {'0': 22, '1': 958, '01': 20}\n",
      "Final population: {'0': 10, '1': 984, '01': 6}\n",
      "Mean 01 proportion after 10 trials: 0.0046\n",
      "Final population: {'0': 9, '1': 989, '01': 2}\n",
      "Final population: {'0': 16, '1': 975, '01': 9}\n",
      "Final population: {'0': 18, '1': 982}\n",
      "Final population: {'0': 4, '1': 994, '01': 2}\n",
      "Final population: {'0': 18, '1': 981, '01': 1}\n",
      "Final population: {'0': 1, '1': 997, '01': 2}\n",
      "Final population: {'1': 980, '0': 18, '01': 2}\n",
      "Final population: {'1': 984, '0': 14, '01': 2}\n",
      "Final population: {'1': 952, '0': 29, '01': 19}\n",
      "Final population: {'1': 988, '0': 11, '01': 1}\n",
      "Mean 01 proportion after 10 trials: 0.004000000000000001\n",
      "Final population: {'1': 977, '0': 21, '01': 2}\n",
      "Final population: {'1': 976, '0': 20, '01': 4}\n",
      "Final population: {'1': 979, '0': 21}\n",
      "Final population: {'1': 979, '0': 12, '01': 9}\n",
      "Final population: {'1': 978, '0': 21, '01': 1}\n",
      "Final population: {'1': 993, '0': 7}\n",
      "Final population: {'1': 979, '0': 21}\n",
      "Final population: {'1': 983, '0': 11, '01': 6}\n",
      "Final population: {'1': 980, '0': 16, '01': 4}\n",
      "Final population: {'1': 971, '0': 25, '01': 4}\n",
      "Mean 01 proportion after 10 trials: 0.003\n",
      "Final population: {'1': 983, '0': 17}\n",
      "Final population: {'1': 978, '0': 18, '01': 4}\n",
      "Final population: {'1': 986, '0': 13, '01': 1}\n",
      "Final population: {'1': 989, '0': 11}\n",
      "Final population: {'0': 12, '1': 977, '01': 11}\n",
      "Final population: {'0': 12, '1': 984, '01': 4}\n",
      "Final population: {'1': 986, '0': 14}\n",
      "Final population: {'1': 976, '0': 23, '01': 1}\n",
      "Final population: {'1': 979, '0': 20, '01': 1}\n",
      "Final population: {'1': 992, '0': 8}\n",
      "Mean 01 proportion after 10 trials: 0.0022\n",
      "Final population: {'1': 981, '0': 19}\n",
      "Final population: {'1': 988, '0': 12}\n",
      "Final population: {'1': 988, '0': 12}\n",
      "Final population: {'1': 987, '0': 11, '01': 2}\n",
      "Final population: {'1': 976, '0': 22, '01': 2}\n",
      "Final population: {'1': 987, '0': 9, '01': 4}\n",
      "Final population: {'1': 971, '0': 23, '01': 6}\n",
      "Final population: {'1': 992, '0': 8}\n",
      "Final population: {'1': 992, '0': 8}\n",
      "Final population: {'1': 987, '0': 13}\n",
      "Mean 01 proportion after 10 trials: 0.0014\n",
      "Final population: {'1': 994, '0': 6}\n",
      "Final population: {'1': 977, '0': 23}\n",
      "Final population: {'1': 986, '0': 11, '01': 3}\n",
      "Final population: {'1': 976, '0': 19, '01': 5}\n",
      "Final population: {'1': 983, '0': 13, '01': 4}\n",
      "Final population: {'1': 984, '0': 14, '01': 2}\n",
      "Final population: {'1': 975, '0': 22, '01': 3}\n",
      "Final population: {'1': 974, '0': 26}\n",
      "Final population: {'1': 986, '0': 12, '01': 2}\n",
      "Final population: {'1': 986, '0': 11, '01': 3}\n",
      "Mean 01 proportion after 10 trials: 0.0022\n",
      "Final population: {'1': 12, '0': 988}\n",
      "Final population: {'1': 18, '0': 978, '01': 4}\n",
      "Final population: {'1': 14, '0': 981, '01': 5}\n",
      "Final population: {'1': 9, '0': 986, '01': 5}\n",
      "Final population: {'1': 11, '0': 989}\n",
      "Final population: {'1': 10, '0': 990}\n",
      "Final population: {'1': 14, '0': 985, '01': 1}\n",
      "Final population: {'1': 22, '0': 976, '01': 2}\n",
      "Final population: {'1': 24, '0': 976}\n",
      "Final population: {'1': 28, '0': 971, '01': 1}\n",
      "Mean 01 proportion after 10 trials: 0.0018000000000000002\n",
      "Final population: {'1': 19, '0': 981}\n",
      "Final population: {'1': 22, '0': 978}\n",
      "Final population: {'1': 21, '0': 976, '01': 3}\n",
      "Final population: {'1': 18, '0': 976, '01': 6}\n",
      "Final population: {'1': 29, '0': 970, '01': 1}\n",
      "Final population: {'1': 33, '0': 966, '01': 1}\n",
      "Final population: {'1': 9, '0': 990, '01': 1}\n",
      "Final population: {'1': 20, '0': 976, '01': 4}\n",
      "Final population: {'1': 29, '0': 965, '01': 6}\n",
      "Final population: {'1': 18, '0': 982}\n",
      "Mean 01 proportion after 10 trials: 0.0022000000000000006\n",
      "Final population: {'1': 12, '0': 988}\n",
      "Final population: {'1': 13, '0': 979, '01': 8}\n",
      "Final population: {'1': 21, '0': 974, '01': 5}\n",
      "Final population: {'1': 16, '0': 973, '01': 11}\n",
      "Final population: {'1': 31, '0': 959, '01': 10}\n",
      "Final population: {'1': 12, '0': 988}\n",
      "Final population: {'1': 16, '0': 980, '01': 4}\n",
      "Final population: {'1': 20, '0': 973, '01': 7}\n",
      "Final population: {'1': 16, '0': 982, '01': 2}\n",
      "Final population: {'1': 26, '0': 969, '01': 5}\n",
      "Mean 01 proportion after 10 trials: 0.005200000000000001\n",
      "Final population: {'1': 981, '0': 19}\n",
      "Final population: {'1': 963, '0': 21, '01': 16}\n",
      "Final population: {'1': 975, '0': 20, '01': 5}\n",
      "Final population: {'1': 978, '0': 22}\n",
      "Final population: {'1': 985, '0': 12, '01': 3}\n",
      "Final population: {'1': 983, '0': 16, '01': 1}\n",
      "Final population: {'1': 990, '0': 10}\n",
      "Final population: {'1': 984, '0': 10, '01': 6}\n",
      "Final population: {'1': 989, '0': 11}\n",
      "Final population: {'1': 963, '0': 36, '01': 1}\n",
      "Mean 01 proportion after 10 trials: 0.0032\n",
      "Final population: {'1': 972, '0': 22, '01': 6}\n",
      "Final population: {'1': 967, '0': 33}\n",
      "Final population: {'1': 986, '0': 12, '01': 2}\n",
      "Final population: {'1': 968, '0': 20, '01': 12}\n",
      "Final population: {'1': 966, '0': 9, '01': 25}\n",
      "Final population: {'1': 985, '0': 7, '01': 8}\n",
      "Final population: {'1': 978, '0': 15, '01': 7}\n",
      "Final population: {'1': 976, '0': 13, '01': 11}\n",
      "Final population: {'0': 19, '1': 978, '01': 3}\n",
      "Final population: {'0': 13, '1': 964, '01': 23}\n",
      "Mean 01 proportion after 10 trials: 0.0097\n",
      "Final population: {'0': 23, '1': 977}\n",
      "Final population: {'0': 19, '1': 974, '01': 7}\n",
      "Final population: {'0': 24, '1': 976}\n",
      "Final population: {'0': 20, '1': 962, '01': 18}\n",
      "Final population: {'0': 17, '1': 983}\n",
      "Final population: {'0': 24, '1': 974, '01': 2}\n",
      "Final population: {'0': 22, '1': 968, '01': 10}\n",
      "Final population: {'1': 963, '0': 22, '01': 15}\n",
      "Final population: {'1': 977, '0': 22, '01': 1}\n",
      "Final population: {'1': 977, '0': 15, '01': 8}\n",
      "Mean 01 proportion after 10 trials: 0.0060999999999999995\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "from pathlib import Path\n",
    "\n",
    "generations = 500\n",
    "trials = 10\n",
    "history = []\n",
    "\n",
    "def time_step():\n",
    "    mutation_step()\n",
    "    fusion_step()\n",
    "    offspring_step()\n",
    "\n",
    "def simulate(regimen):\n",
    "    clone_pop = dict(pop)\n",
    "    history.append(clone_pop)\n",
    "    for i in range(generations):\n",
    "        drug_step(i, regimen)\n",
    "        time_step()\n",
    "        clone_pop = dict(pop)\n",
    "        history.append(clone_pop)\n",
    "\n",
    "fractions = [0.9, 0.5, 0.1]\n",
    "for regimen_num in range(len(regimens)):\n",
    "    for fraction in fractions:\n",
    "        fraction_base = fraction\n",
    "\n",
    "        final_pops = []\n",
    "        fused_01_outcomes = []\n",
    "        fused_11_outcomes = []\n",
    "        for i in range(trials):\n",
    "            simulate(regimen_num)\n",
    "            \n",
    "            print('Final population: {}'.format(pop))\n",
    "            final_pops.append([val for pair in zip(pop.keys(), pop.values()) for val in pair])\n",
    "            \n",
    "            fused_01 = 0\n",
    "            fused_11 = 0\n",
    "            for (hap, count) in pop.items():\n",
    "                if hap == \"01\":\n",
    "                    fused_01 += count\n",
    "            fused_01_proportion = fused_01 / float(pop_size)\n",
    "            fused_01_outcomes.append(fused_01_proportion)\n",
    "\n",
    "        # Calculate average proportion over all trials\n",
    "        avg_01_proportion = 0.0\n",
    "        for proportion in fused_01_outcomes:\n",
    "            avg_01_proportion += proportion\n",
    "        avg_01_proportion = avg_01_proportion / len(fused_01_outcomes)\n",
    "        print('Mean 01 proportion after {} trials: {}'.format(trials, avg_01_proportion))\n",
    "\n",
    "        filename = \"r{}_output_{}.csv\".format(regimen_num, fraction_base)\n",
    "        filepath = Path('./output') / filename\n",
    "        with filepath.open('w') as csvfile:\n",
    "            csvwriter = csv.writer(csvfile)\n",
    "            csvwriter.writerow([fraction_base])\n",
    "            csvwriter.writerow([avg_01_proportion])\n",
    "            csvwriter.writerow([regimen_names[regimen_num]])\n",
    "            csvwriter.writerows(final_pops)\n",
    "\n",
    "# plt.figure(num=None, figsize=(14, 14), dpi=80, facecolor='w', edgecolor='k')\n",
    "# plt.subplot2grid((2,2), (0,0), colspan=2)\n",
    "# stacked_trajectory_plot()\n",
    "# #plt.subplot2grid((3,2), (1,0), colspan=2)\n",
    "# #snp_trajectory_plot()\n",
    "# plt.subplot2grid((2,2), (1,0))\n",
    "# diversity_plot()\n",
    "# plt.subplot2grid((2,2), (1,1))\n",
    "# divergence_plot()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plotting\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Plotting  drug_off\n",
      "Plotting  drug_off\n",
      "Plotting  drug_off\n",
      "Plotting  drug_on\n",
      "Plotting  drug_on\n",
      "Plotting  drug_on\n",
      "Plotting  drug_off_on\n",
      "Plotting  drug_off_on\n",
      "Plotting  drug_off_on\n",
      "Plotting  drug_on_off\n",
      "Plotting  drug_on_off\n",
      "Plotting  drug_on_off\n",
      "Plotting  drug_off_on_off_on\n",
      "Plotting  drug_off_on_off_on\n",
      "Plotting  drug_off_on_off_on\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAKIAAAHNCAYAAACdJ6AgAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAxOAAAMTgF/d4wjAAAdWUlEQVR4nO3dfXBV9b3v8fc3hKfaUKsWJcaQtgZTDMjDjXKxRfGeQhSMIqAMD51UUhKdDrd32iOO9lBbrs6heua2tqWgCFJxaLV4uURLe9V6cDi3WCgNohQJU0IQibQ+IQglke/9Yy9yAoHsHfbe5Ifr85rZw17rt/ZvfTd8+K2nvdc2d0ekq+V0dQEioCBKIBRECYKCKEFQECUICqIEIberVmxmzTk5OR901frlzDp69Gieu/c8VXuXBTEnJ+eDcePG3d5V65czq7a2dkFH7do0SxAURAmCgihBUBAlCAqiBEFBlCAoiBIEBVGCoCBKEBRECYKCKEFQECUICqIEQUGUICiIEgQFUYKgIEoQFEQJgoIoQVAQJQgKogRBQZQgKIgSBAVRgqAgShAURAmCgihBUBAlCAqiBEFBlCAoiBIEBVGCoCBKEBRECYKCKEFQECUICqIEQUGUICiIEgQFUYKgIEoQFEQJgoIoQVAQJQgKogThtINoZr/PZCESbx3+TK6Z9emguTjDtUiMJfu95vcBB6zNvGPTnqWaJIaSbZr3Ahe6e06bRzd3zwHeOgP1SUwkC+L/Awafom1zhmuRGOtw0+zukztoG5/5ciSudPpGgpDO6Zv7MliHxFw6I+LejFUhsXfaQXT3RZksROIt2XlEzOyzwASgMJrVCKxy93ezWZjES4cjoplNBLYBY4De0WMssDVqE8mIZCPi/cBV7t7QdqaZfR5YA6zMUl0SM8n2EbudGEIAd99JCpt1kVQlC+IGM1tiZlea2YXR40ozWwJsPBMFSjwkC+JMYCewBKiPHo8Bu4Dbs1uaxEmyS3yHgHnRQyRrYrOft3r16v/T1TWEpqKi4qauruEYXWuWIMQuiPX19YwcOZIBAwZQVlbG66+/3to2e/ZsioqKMDPq6upS7nPfvn2Ul5dTXFxMaWkpL7/88imXffbZZykpKaG4uJhbbrmF/fv3p9TW2NjIjTfeyGWXXcbAgQP5yU9+AsCBAwcYO3YsF1xwAeeee26n1heS2AWxurqaWbNmsX37dubMmUNlZWVr26RJk1i3bh39+/fvVJ933303I0aMoL6+nqVLlzJ16lSam5vbLXfgwAFmzpzJqlWrqK+vJz8/n3nz5iVtc3cmTJjA1772Nd544w22bt3KrbfeCkD37t2ZM2cOL7zwQqfWF5qUg2hmszqaPhvs27ePjRs3Mn36dAAmTpzI7t272bFjBwCjRo2ioKCg0/0+9dRT1NTUAFBWVkZ+fj5r165tt9yaNWsYOnQoJSUlANx5552sWLEiaduLL75Iz549mTz5Pz8eeuGFFwLQs2dPrrvuupOOhh31GZrOjIgXJ5kO3u7du+nXrx+5uYljNDOjsLCQxsbGpK+tqqpi9erV7ea/8847NDc3c9FFF7XOKyoqOmmfjY2Nx422RUVF7N27l5aWlg7btm7dyuc+9zmmTJnC0KFDmTBhAn/961+T1txRn6FJ+ajZ3b/X0fQn3eLFi7ts3S0tLfz+979n/fr1XH755SxcuJBbb72VjRs/OdcUOrNpnmhm90TPLzazQdkrKzsuueSS40YEd6exsZHCwsIkrzy1888/n9zcXJqamlrnNTQ0nLTPwsJCdu3addxyx0boZG1Dhw7l8ssvB2DGjBls2rTppPuhqa4vNCkF0cx+AFQBldGso8BZ93nEvn37MmzYMJYvXw7AypUrKSgo4NJLL02r38mTJ7Nw4UIANmzYwJ49e7jmmmvaLVdeXs6mTZvYtm0bAAsWLGDKlClJ266//nrefPNN9uzZA8BvfvMbvvSlL9G9e/cO6+qoz9Ck+l/jJmAY0fVld99rZp/OWlVZtGjRIiorK3nggQfo06cPS5cubW2rrq7mueeeo6mpibFjx5KXl9d6IFNVVUVFRQUVFRXt+pw/fz4zZsyguLiYHj16sHz58taQzJ07l/z8fGpqasjLy2Px4sXcfPPNtLS0UFpayrJlywA6bDvnnHNYuHAh48aNw935zGc+wy9/+cvW9Q8ePJi//e1v7N+/n4KCAkaPHs0TTzzRYZ+hMffk35M3s/XuPsLM/uzuQ6N5W9z9tDfP3bp1+/u4cePO2PVqXVlp70xeWamtrV3g7qc8JZHqiLjLzL4CuJl1B+4B6jJQnwiQehBnA8uAQcBB4CVgWraKkvhJKYju/jZQbmafIrE5P5jdsjIvpAv80l6qR803mlkfd/8IuMPMfm1mpVmuTWIk1fOI97v7fjO7ApgOPA/8PHtlSdykGsRj14TGAI9E32k+JzslSRylGsRuZnYVMJHEgQpAx2dTRToh1SB+l8SVlHXu/hczuwzYnr2yJG5SPWquBWrbTL9BYnQUyYiUr36b2ZXAEKDXsXnu/nAWapIYSimI0aduJpG4/81a4KvAi4CCKBmR6j7iVGAk8Ka7TwTKSHwCRyQjUg3iYXc/DOSYmUX7iF/MYl0SM6nuIx6KPuxQBzxkZm8C3bJWlcROqiPiHUAP4NtAH+BqYEa2ipL4SfX0zWvR04PAN7JXjsRVsht19jazO8xsipnlmNlDZrbFzJ42s7PuW3wSrmSb5keBccAs4HfAucBdQAOwMJuFSbwk2zQPc/eBZtaLxK8IjHX3o8AaM9uS/fIkLpKNiP8AiE7dNEQhPKbj7zKKdEKyEbFX9P1lO+E5JG7sLpIRyYLYG2h7n422z/UzuZIxye4YW3SG6pCYi91t6SRMCqIEQUGUICiIEgQFUYKgIEoQFEQJgoIoQVAQJQgKogRBQZQgKIgSBAVRgqAgShAURAmCgihBUBAlCAqiBEFBlCAoiBIEBVGCoCBKEBRECYKCKEFQECUICqIEQUGUICiIEgQFUYKgIEoQFEQJgoIoQVAQJQgKogRBQZQgKIgSBAVRgqAgShAURAmCgihBUBAlCAqiBEFBlCCYe9f8tqOZeW5urn5YMiZaWlrM3e1U7cl+nTRrcnNz919//fUPdtX65cyqra2t6ahdm2YJgoIoQVAQJQgKogRBQZQgKIgSBAVRgqAgShAURAmCgihBUBAlCAqiBEFBlCAoiBIEBVGCoCBKEBRECYKCKEFQECUICqIEQUGUICiIEgQFUYKgIEoQFEQJgoIoQVAQJQgKogRBQZQgKIgSBAVRgqAgShAURAmCgihBUBAlCAqiBEFBlCAoiBIEBVGCoCBKEBRECYKCKEFQECUICqIEQUGUIHQYRDO7us3zT5nZz8xss5k9bmbnZr06iY1kI+JP2jz/AZAHVAPvAT/OVlESP8l+r7ntDz3/N2Ckux8ysz8Cm7NXlsRNsiDmmFkeiUC2uPshAHc/amYtWa9OYiNZEAcB75MIoptZgbu/aWa90YGOZFCHQXT3U4WtJ4l9RZGMSDYinpS7vw+sz2wpEmenvXk1s/syWIfEXDr7eXszVoXE3mkH0d0XZbIQibek+4hm9llgAlAYzWoEVrn7u9ksTOIl2SW+icA2YAzQO3qMBbZGbSIZkWxEvB+4yt0b2s40s88Da4CVWapLYibZPmK3E0MI4O47Oc1TPyInkyyIG8xsiZldaWYXRo8rzWwJsPFMFCjxkCyIM4GdwBKgPno8BuwCbs9uaRInyS7xHQLmRQ+RrInNft7q1av1n+kEFRUV/9LVNRyjT9BIEGIXxPr6ekaOHMmAAQMoKyvj9ddfb22bPXs2RUVFmBl1dXUp97lv3z7Ky8spLi6mtLSUl19++ZTLPvvss5SUlFBcXMwtt9zC/v37U2p78MEHKS0tZeDAgUyYMIH3338fgC1btjBkyJDWR1FREeedd15K7zcksQtidXU1s2bNYvv27cyZM4fKysrWtkmTJrFu3Tr69+/fqT7vvvtuRowYQX19PUuXLmXq1Kk0Nze3W+7AgQPMnDmTVatWUV9fT35+PvPmzUva9vzzz7N06VL+8Ic/sHXrVoYPH869994LwKBBg6irq2t9jB8/nmnTpqX0fkOSchDNbFZH02eDffv2sXHjRqZPnw7AxIkT2b17Nzt27ABg1KhRFBQUdLrfp556ipqaGgDKysrIz89n7dq17ZZbs2YNQ4cOpaSkBIA777yTFStWJG3bvHkzX/7yl8nLywPghhtu4IknnmjX/+HDh3nyySeZOXNmSu83JJ0ZES9OMh283bt3069fP3JzE8doZkZhYSGNjY1JX1tVVcXq1avbzX/nnXdobm7moosuap1XVFR00j4bGxuPG22LiorYu3cvLS0tHbYNHz6cF154gaamJtydJ598kg8//JB33z3+cv8zzzzDF77wBYYMGZL2+z3TUj5qdvfvdTT9Sbd48eIuW/fo0aP5zne+w/jx4+nWrRsTJkwAaA3YMY899ljraHi26cymeaKZ3RM9v9jMBmWvrOy45JJLWkcZAHensbGRwsLCJK88tfPPP5/c3Fyamppa5zU0NJy0z8LCQnbt2nXccsdGrI7aILGp3rhxI6+88grXXnstBQUF9OnTp3X5nTt3sn79eqZOnZrV95stKQXRzH4AVAGV0ayjwFn3ecS+ffsybNgwli9fDsDKlSspKCjg0ksvTavfyZMns3DhQgA2bNjAnj17uOaaa9otV15ezqZNm9i2bRsACxYsYMqUKUnbAPbuTXwO+aOPPmLu3Lncddddx/W9ZMkSJkyYwLnnnpv195sNqW6abwKGEV1fdve9ZvbprFWVRYsWLaKyspIHHniAPn36sHTp0ta26upqnnvuOZqamhg7dix5eXmtO/ZVVVVUVFRQUVHRrs/58+czY8YMiouL6dGjB8uXL6d79+4AzJ07l/z8fGpqasjLy2Px4sXcfPPNtLS0UFpayrJlywA6bAMYM2YMR48e5ciRI8yYMYNvfvObrW1Hjx7l8ccf5xe/+EWn3m9IzN2TL2S23t1HmNmf3X1oNG+Lu5/25rl79+4fXH/99Q+e7us7S1dW2juTV1Zqa2tr3P2UpyRSHRF3mdlXSHy3uTtwD1CXgfpEgNSDOBtYRuIL9weBl4BpHb5CpBNSCqK7vw2Um9mnSGzOD2a3rMwL6QK/tJfqUfONZtbH3T8C7jCzX5tZaZZrkxhJ9Tzi/e6+38yuAKYDzwM/z15ZEjepBvHYnb/GAI9E32k+JzslSRylGsRuZnYVMJHEgQpA9+yUJHGUahC/S+JKyjp3/4uZXQZsz15ZEjepHjXXArVtpt8gMTqKZETKn74xsyuBIUCvY/Pc/eEs1CQxlFIQo0/dTCJx/5u1wFeBFwEFUTIi1X3EqcBI4E13nwiUkfgEjkhGpBrEw+5+mMTN3S3aR/xiFuuSmEl1H/FQ9GGHOuAhM3sT6Ja1qiR2Uh0R7wB6AN8G+gBXAzOyVZTET6qnb16Lnh4EvpG9ciSukt2os7eZ3WFmU8wsx8weMrMtZva0mZ113+KTcCXbND8KjANmAb8DzgXuAhqAhdksTOIl2aZ5mLsPNLNeJH5FYKy7HwXWmNmW7JcncZFsRPwHQHTqpiEK4THt76khcpqSjYi9ou8v2wnPIXFjd5GMSBbE3kDb+2y0fZ78638iKUp2x9iiM1SHxFzsbksnYVIQJQgKogRBQZQgKIgSBAVRgqAgShAURAmCgihBUBAlCAqiBEFBlCAoiBIEBVGCoCBKEBRECYKCKEFQECUICqIEQUGUICiIEgQFUYKgIEoQFEQJgoIoQVAQJQgKogRBQZQgKIgSBAVRgqAgShAURAmCgihBUBAlCAqiBEFBlCAoiBIEBVGCoCBKEBRECYKCKEFQECUICqIEwdy75rcdzczN7GjyJeWTwN1z3N1O1Z7s10mzxszeHj9+/OiuWr+cWbW1tc931K5NswRBQZQgKIgSBAVRgqAgShAURAmCgihBUBAlCAqiBEFBlCAoiBIEBVGCoCBKEBRECYKCKEFQECUICqIEQUGUICiIEgQFUYKgIEoQFEQJgoIoQVAQJQgKogRBQZQgKIgSBAVRgqAgShAURAmCgihBUBAlCAqiBEFBlCAoiBIEBVGCoCBKEBRECYKCKEFQECUICqIEQUGUICiIEgQFUYKQkR+FNLPPAv2BFmCHux/ORL8SH2mNiGZWYGbPAn8H/gSsBd41sx+aWfdMFCjxkO6meSmwArgA+B/Aw8Dngb7Ag2n2LTGSbhAvdPcn3f09d38YGOfubwMzgXHplydxkW4Qj5pZXwAzuxT4GMDdPwaOpNm3xEi6Byv/Bmw2s83AMKAGwMwuAnal2bfESFpBdPcnzOyPQCmw2d13RPObgBsyUJ/ERNqnb9z9DeCNDNQiMZa1E9pmdl+2+pZPnmxeWdmbxb7lEyZrQXT3RdnqWz550t5HjC7vTQAKo1mNwCp3fzfdviU+0r3ENxHYBowBekePscDWqE0kJemOiPcDV7l7Q9uZZvZ5YA2wMs3+JSbS3UfsdmIIAdx9Jxn6ZI/EQ7pB3GBmS8zsSjO7MHpcaWZLgI2ZKFDiId0gzgR2AkuA+ujxGInLe7en2bfESLqX+A4B86KHyGmLzX7c6tWrt3Z1DaGpqKgY2NU1HKPvrEgQYhfE+vp6Ro4cyYABAygrK+P1119vbZs9ezZFRUWYGXV1dSn3uW/fPsrLyykuLqa0tJSXX375lMs+++yzlJSUUFxczC233ML+/fuTtu3cuZPhw4czZMgQSktLmTx5Mu+99x4AW7ZsYdSoUZSUlFBaWsrtt9/OoUOHWvt85ZVXuOKKKxgwYADXXXcde/bsSfl9nUmxC2J1dTWzZs1i+/btzJkzh8rKyta2SZMmsW7dOvr379+pPu+++25GjBhBfX09S5cuZerUqTQ3N7db7sCBA8ycOZNVq1ZRX19Pfn4+8+bNS9qWn5/PunXrqKur47XXXiM/P5/77rsPgF69evHTn/6Ubdu2sXnzZg4ePMj8+fMBOHr0KNOmTeNHP/oR27dv54YbbuBb3/pW5//SzoCMBdHMZnU0HYJ9+/axceNGpk+fDsDEiRPZvXs3O3bsAGDUqFEUFBR0ut+nnnqKmpoaAMrKysjPz2ft2rXtlluzZg1Dhw6lpKQEgDvvvJMVK1YkbevZsye9e/cG4OOPP+bgwYOYGQDFxcUMHjwYgG7dulFWVkZDQwMAf/rTn8jNzWX06NFA4j9hbW0thw+H9yXLTI6IFyeZ7nK7d++mX79+5OYmjtHMjMLCQhobG5O+tqqqitWrV7eb/84779Dc3MxFF13UOq+oqOikfTY2Nh432hYVFbF3715aWlo6bAM4cuQIQ4YM4YILLqC+vp7vf//77fo/ePAgixcv5qabbjrp+vLy8ujTpw9vvfVW0vd7pmUsiO7+vY6mz3aLFy+moqKiy9bfo0cP6urqePvttykpKWHRouM/3HTkyBFuu+02xowZw4QJE7qoytOXyU3zRDO7J3p+sZkNylTfmXLJJZccN8q4O42NjRQWFiZ55amdf/755Obm0tTU1DqvoaHhpH0WFhaya9eu45Y7NkJ31NZWjx49+PrXv84TTzzROq+5uZnbbruNfv368eMf//iU6/vwww/54IMPyM/PP+33my0ZCaKZ/QCoAiqjWUeB4D6P2LdvX4YNG8by5csBWLlyJQUFBVx66aVp9Tt58mQWLlwIwIYNG9izZw/XXHNNu+XKy8vZtGkT27ZtA2DBggVMmTIladuuXbv46KOPgMQByNNPP926X9jS0sKUKVM477zzeOSRR1r3HQGGDx9Oc3MzL730EgCLFi3ixhtvpFevXmm932zI1Antm0h8i28jgLvvNbNPZ6jvjFq0aBGVlZU88MAD9OnTh6VLl7a2VVdX89xzz9HU1MTYsWPJy8trPZCpqqqioqLipJvn+fPnM2PGDIqLi+nRowfLly+ne/fEjS7mzp1Lfn4+NTU15OXlsXjxYm6++WZaWlooLS1l2bJlAB22vfrqq9x7771AIojDhg3j4YcfBuBXv/oVzzzzDIMHD2bo0KEAXH311fzsZz8jJyeH5cuXU11dzeHDh8nPzz9uJA2JuXv6nZitd/cRZvZndx8azdvi7qfcPOfk5DSNHz9+dNorT5GurLR3Jq+s1NbWPu/upzwlkakRcZeZfQXw6J439wB1GepbYiBTQZwNLAMGAQeBl4BpGepbYiAjQYzud1NuZp8isbk/mIl+MymkC/zSXqaOmm80sz7u/hFwh5n92sxKM9G3xEOmziPe7+77zewKYDrwPPDzDPUtMZCpILZEf44BHom+03xOhvqWGMhUELuZ2VXARBIHKgC6Y6ykLFNB/C6JKynr3P0vZnYZsD1DfUsMZOqouRaobTP9BonRUSQlGfvOipldCQwBWi9kRrczFkkqUz9vcQ8wicT9b9YCXwVeJHFzd5GkMrWPOBUYCbzp7hOBMhKfwBFJSaaCeDj6kZ8cM7NoH/GLGepbYiBT+4iHog871AEPmdmbQLcM9S0xkKkR8Q6gB/BtoA9wNTAjQ31LDGTq9M1r0dODwDcy0afES7o36uxtZneY2RQzyzGzh8xsi5k9bWbBfYtPwpXupvlREj91Ngv4HXAucBfQACxMs2+JkXQ3zcPcfaCZ9SLxKwJj3f0osMbMtqRfnsRFuiPiPwCiUzcNUQiPaX/PDZFTSHdE7BV9f9lOeA6JG7uLpCTdIPYG2t6Ho+3z9L8eKLGR7h1jizJUh8Rc7G5LJ2FSECUICqIEQUGUICiIEgQFUYKgIEoQFEQJgoIoQVAQJQgKogRBQZQgKIgSBAVRgqAgShAURAmCgihBUBAlCAqiBEFBlCAoiBIEBVGCoCBKEBRECYKCKEFQECUICqIEQUGUICiIEgQFUYKgIEoQFEQJgoIoQVAQJQgKogRBQZQgKIgSBAVRgqAgShAURAmCgihBUBAlCAqiBMHcu+a3G83sH8DfumTl0hU+5+49T9XYZUEUaUubZgmCgihBUBAlCAqiBEFBlCAoiBIEBVGCoCBKEBRECYKCKEFQECUICqIE4awMopn93cyKuriG88zsP8yszszuPXG6K2uL6vuimW0ysz+b2ddPnO7q+k6U29UFZIqZ5bp7yxlc5VeBA+5+dbT+29pOB2ASsMHdqwHMbE7b6eC4e/APoAL4C/Aq8EPg70AR0ADMB/4IPAlcC9S1eV0p0NBmuhrYDmwC/iXx9jtcbzfgQeC16PEToAfwT0Aj8AFQd7LpDvocG63/VWAtMDCaf220jgXAZuB14L8kqe/TwJI29X0vmv81oInE5z3rgLknTA/soM8ZUW2vAs8BF0fzK4EXgBXAFmAj8IWM/Rt3dchSCGFf4J02/2CzAG8TxMX85+cqTxnE6Ple4KJo+vspBPEO4N+BniS2Hr8B5rT5h1nVZtnjppO8l0HR9DRgK2BR7S3AVVFbDfC7JP3NJ/EfMAc4B/gzcFvUdh/wozbLHjd9iv5Ko8AeC9+9wJo27+8D4PPR9L8CizL173w27COOAF51963R9GPAkTbtj3v0N5PEdcBv3b0pmn40hdf8U9T/Pzyx2X+UxCb5dF0FbHH3LQDu/iSQD1wcte9w91ei538AvphCfY+6+1F3Pwj8Is36RpP4O9oTTS8ArjOzbsdqcvednagvZWdDEE90YugOtHneQmJzekyvTvRzOuvOtMNtnn9M5/fhM13fif2lW98pnQ1B/AMw2MxKounbSeynncxfgf5m9rloekabtpeAsWbWN5qemcK6XwC+ZmY9zCwXqAL+b6eqP956YJCZlQKY2RRgT/Q4HS8AMy3hHBLvN536XgLKzSw/mq4BXnT3j9PoMyXBHzW7+9/M7Hbgf5vZEeC3JPazTrbsW2b2Q+CPZvY2sKZN2xYz+5/Af5jZh1E/HyRZ/SMkNj+boul/B36U5nuZBvwiCvZ7wGR3dzM7nS7nAQ+TOHgAeNrdn0qjvtfM7J+B30b17Aa+cbr9dUasvjxlZnnu/mH0/L8D5e5+fReXJZwFI2KG/auZXQ10B94icTpHAhCrEfFkon3Gk+1XPe/u/3yafS4kcbR/ov/q7oc62dcQ4PGTNC1z9//V+erAzFYDhSfMfs/dR59Of5kQ+yBKGM6Go2aJAQVRgqAgShAURAmCgihB+P9PkVc54EkOTAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 800x560 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "\n",
    "plt.figure(figsize=(10, 7), dpi=80)\n",
    "gs = gridspec.GridSpec(3, 5)\n",
    "gs.update(wspace=0, hspace=0) # set the spacing between axes. \n",
    "\n",
    "\n",
    "results = [file for file in Path('./output').iterdir() if file.is_file() and file.name.startswith('r')]\n",
    "results.sort()\n",
    "for count, file in enumerate(results):\n",
    "   row_num = count % 3\n",
    "   col_num = int(count) / 3\n",
    "   \n",
    "   with file.open('r') as csvfile:\n",
    "      reader = csv.reader(csvfile)\n",
    "      rows = [row for row in reader]\n",
    "      starting_fraction = float(rows[0][0])\n",
    "      proportions = [float(val) for val in rows[1]]\n",
    "      proportion_fused = sum(proportions)\n",
    "\n",
    "      regimen_name = rows[2][0]\n",
    "      print('Plotting ', regimen_name)\n",
    "\n",
    "      ax = plt.subplot(gs[row_num, col_num])\n",
    "      ax.set_facecolor((0.3, 0.3, 0.3, proportion_fused / 0.02))\n",
    "      ax.text(0.5, 0.5, '01: {:.5f}'.format(proportions[0]),\n",
    "         bbox={'facecolor':'white','alpha':1,'edgecolor':'none','pad':1},\n",
    "         ha='center', va='center')\n",
    "\n",
    "      # Add the drug regimen to the bottom row label.\n",
    "      if row_num == 2 or col_num == 0:\n",
    "         if row_num == 2:\n",
    "            ax.set_xlabel(regimen_name)\n",
    "            ax.xaxis.set_ticklabels([])\n",
    "            ax.tick_params(axis='both', which='both', bottom=False, left=False, labelleft=False)\n",
    "         if col_num == 0:\n",
    "            # Add the proportion to the first column label\n",
    "            ax.set_ylabel('Base: {}'.format(starting_fraction))\n",
    "            ax.yaxis.set_ticklabels([])\n",
    "            ax.xaxis.set_ticklabels([])\n",
    "            ax.tick_params(axis='both', which='both', bottom=False, left=False)\n",
    "      else:\n",
    "         ax.tick_params(axis='both', which='both', bottom=False, labelbottom=False, left=False, labelleft=False)\n",
    "\n",
    "   row_num += 1\n",
    "   col_num += 1\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
  },
  "kernelspec": {
   "display_name": "Python 3.9.10 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
