{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "try:\n",
    "    import itertools.izip as zip\n",
    "except ImportError:\n",
    "    import itertools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_size = 1000\n",
    "alphabet = ['0', '1']\n",
    "seq_length = 1\n",
    "\n",
    "base_haplotype = \"0\"\n",
    "fraction_base = 0.5\n",
    "\n",
    "fitness_nodrug = {'base': 1, 'resistant': 0.8, 'fusion': 0.9}\n",
    "fitness_drug = {'base': 0.8, 'resistant': 1.0, 'fusion': 0.9}\n",
    "\n",
    "fitness_base = fitness_nodrug['base']\n",
    "fitness_resistant = fitness_nodrug['resistant']\n",
    "fitness_fusion = fitness_nodrug['fusion']\n",
    "\n",
    "fitness_effect = 1 # fitness effect if a functional mutation occurs\n",
    "fitness_chance = 0.01 # chance that a mutation has a fitness effect\n",
    "\n",
    "mutation_rate = 0.005\n",
    "fusion_rate = 0.01\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop = {}\n",
    "pop[\"0\"] = pop_size * fraction_base\n",
    "pop[\"1\"] = pop_size * (1 - fraction_base)\n",
    "\n",
    "print(\"Starting population: {}\".format(pop))\n",
    "\n",
    "fitness = {}\n",
    "fitness[\"0\"] = fitness_base\n",
    "fitness[\"1\"] = fitness_resistant\n",
    "\n",
    "print(\"Fitness: {}\".format(fitness))\n",
    "\n",
    "def get_random_haplotype():\n",
    "    \"\"\"Returns a random haplotype from the population, weighted by the frequency of the haplotype\"\"\"\n",
    "    haplotypes = list(pop.keys()) \n",
    "    frequencies = [x/float(pop_size) for x in pop.values()]\n",
    "    total = sum(frequencies)\n",
    "    frequencies = [x / total for x in frequencies]\n",
    "    return np.random.choice(haplotypes, p=frequencies)\n",
    "\n",
    "def update_fitness(haplotype):\n",
    "    old_fitness = fitness[haplotype]\n",
    "    if (np.random.random() < fitness_chance):\n",
    "        return old_fitness * fitness_effect\n",
    "    else:\n",
    "        return old_fitness\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mutation_count():\n",
    "    mean = mutation_rate * pop_size * seq_length\n",
    "    return np.random.poisson(mean)\n",
    "\n",
    "def make_mutant(haplotype):\n",
    "    site = np.random.randint(seq_length)\n",
    "    possible_mutations = list(alphabet)\n",
    "    possible_mutations.remove(haplotype[site])\n",
    "    mutation = np.random.choice(possible_mutations)\n",
    "    new_haplotype = haplotype[:site] + mutation + haplotype[site+1:]\n",
    "    return new_haplotype    \n",
    "\n",
    "def get_mutant_fitness(haplotype):\n",
    "    old_fitness = fitness[haplotype]\n",
    "    if (np.random.random() < fitness_chance):\n",
    "        return old_fitness * fitness_effect\n",
    "    else:\n",
    "        return old_fitness\n",
    "\n",
    "def mutation_event():\n",
    "    haplotype = get_random_haplotype()\n",
    "    \n",
    "    # Only let non-fused cells mutate.\n",
    "    if len(haplotype) > 1:\n",
    "        return\n",
    "\n",
    "    if pop[haplotype] > 1:\n",
    "        pop[haplotype] -= 1\n",
    "        new_haplotype = make_mutant(haplotype)\n",
    "        if new_haplotype in pop:\n",
    "            pop[new_haplotype] += 1\n",
    "        else:\n",
    "            pop[new_haplotype] = 1\n",
    "        if new_haplotype not in fitness:\n",
    "            fitness[new_haplotype] = get_mutant_fitness(haplotype)\n",
    "\n",
    "def mutation_step():\n",
    "    mutation_count = get_mutation_count()\n",
    "    for i in range(mutation_count):\n",
    "        mutation_event()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fusion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fusion_count():\n",
    "    mean = fusion_rate * pop_size\n",
    "    return np.random.poisson(mean)\n",
    "\n",
    "def make_fusion(hap1, hap2):\n",
    "    fused = hap1 + hap2\n",
    "    return ''.join(sorted(fused))\n",
    "\n",
    "def get_fusion_fitness(haplotype):\n",
    "    from math import e\n",
    "    \n",
    "    length = len(haplotype)\n",
    "\n",
    "    # Don't let the haplotype go past 2 chars.\n",
    "    if length > 2:\n",
    "        return 0\n",
    "\n",
    "    # Use a Gaussian function such that the fitness is lower when very short and very long.\n",
    "    # f(x) = ae^(-(x-b)^2/2c^2)\n",
    "    a = fitness_fusion\n",
    "    b = 2 # max fitness at length 2\n",
    "    c = 1 # stdev\n",
    "    \n",
    "    fitness = a * e ** (-1 * ((length - b) ** 2) / ((2 * c) ** 2))\n",
    "    return fitness\n",
    "\n",
    "def fusion_event():\n",
    "    hap1 = get_random_haplotype()\n",
    "    hap2 = get_random_haplotype()\n",
    "\n",
    "    # Don't let the same haplotypes fuse.\n",
    "    if hap1 == hap2:\n",
    "        return\n",
    "        \n",
    "    # If fusion would remove either population of cells, don't let it happen.\n",
    "    if pop[hap1] == 1 or pop[hap2] == 1:\n",
    "        return\n",
    "\n",
    "    pop[hap1] -= 1\n",
    "    pop[hap2] -= 1\n",
    "\n",
    "    new_haplotype = make_fusion(hap1, hap2)\n",
    "\n",
    "    if new_haplotype in pop:\n",
    "        pop[new_haplotype] += 1\n",
    "    else:\n",
    "        pop[new_haplotype] = 1\n",
    "    if new_haplotype not in fitness:\n",
    "        fitness[new_haplotype] = get_fusion_fitness(new_haplotype)\n",
    "\n",
    "def fusion_step():\n",
    "    fusion_count = get_fusion_count()\n",
    "    for i in range(fusion_count):\n",
    "        fusion_event()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Offspring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_offspring_counts():\n",
    "    haplotypes = list(pop.keys())\n",
    "    frequencies = [pop[haplotype]/float(pop_size) for haplotype in haplotypes]\n",
    "    fitnesses = [fitness[haplotype] for haplotype in haplotypes]\n",
    "    weights = [x * y for x,y in zip(frequencies, fitnesses)]\n",
    "    total = sum(weights)\n",
    "    weights = [x / total for x in weights]\n",
    "    return list(np.random.multinomial(pop_size, weights))\n",
    "\n",
    "def offspring_step():\n",
    "    haplotypes = list(pop.keys())\n",
    "    counts = get_offspring_counts()\n",
    "    for (haplotype, count) in zip(haplotypes, counts):\n",
    "        if (count > 0):\n",
    "            pop[haplotype] = count\n",
    "        else:\n",
    "            del pop[haplotype]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Levenshtein import distance as levenshtein_distance\n",
    "\n",
    "def get_distance(seq_a, seq_b):\n",
    "    return levenshtein_distance(seq_a, seq_b)\n",
    "\n",
    "def get_diversity(population):\n",
    "    haplotypes = list(population.keys())\n",
    "    haplotype_count = len(haplotypes)\n",
    "    diversity = 0\n",
    "    for i in range(haplotype_count):\n",
    "        for j in range(haplotype_count):\n",
    "            haplotype_a = haplotypes[i]\n",
    "            haplotype_b = haplotypes[j]\n",
    "            frequency_a = population[haplotype_a] / float(pop_size)\n",
    "            frequency_b = population[haplotype_b] / float(pop_size)\n",
    "            frequency_pair = frequency_a * frequency_b\n",
    "            diversity += frequency_pair * get_distance(haplotype_a, haplotype_b)\n",
    "    return diversity\n",
    "\n",
    "def get_diversity_trajectory():\n",
    "    trajectory = [get_diversity(generation) for generation in history]\n",
    "    return trajectory\n",
    "\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "\n",
    "def diversity_plot():\n",
    "    mpl.rcParams['font.size']=14\n",
    "    trajectory = get_diversity_trajectory()\n",
    "    plt.plot(trajectory, \"#447CCD\")    \n",
    "    plt.ylabel(\"diversity\")\n",
    "    plt.xlabel(\"generation\")\n",
    "\n",
    "def get_divergence(population):\n",
    "    haplotypes = population.keys()\n",
    "    divergence = 0\n",
    "    for haplotype in haplotypes:\n",
    "        frequency = population[haplotype] / float(pop_size)\n",
    "        divergence += frequency * get_distance(base_haplotype, haplotype)\n",
    "    return divergence\n",
    "\n",
    "def get_divergence_trajectory():\n",
    "    trajectory = [get_divergence(generation) for generation in history]\n",
    "    return trajectory\n",
    "\n",
    "def divergence_plot():\n",
    "    mpl.rcParams['font.size']=14\n",
    "    trajectory = get_divergence_trajectory()\n",
    "    plt.plot(trajectory, \"#447CCD\")\n",
    "    plt.ylabel(\"divergence\")\n",
    "    plt.xlabel(\"generation\") \n",
    "\n",
    "def get_frequency(haplotype, generation):\n",
    "    pop_at_generation = history[generation]\n",
    "    if haplotype in pop_at_generation:\n",
    "        return pop_at_generation[haplotype]/float(pop_size)\n",
    "    else:\n",
    "        return 0\n",
    "    \n",
    "def get_trajectory(haplotype):\n",
    "    trajectory = [get_frequency(haplotype, gen) for gen in range(generations)]\n",
    "    return trajectory\n",
    "\n",
    "def get_all_haplotypes():\n",
    "    haplotypes = set()   \n",
    "    for generation in history:\n",
    "        for haplotype in generation:\n",
    "            haplotypes.add(haplotype)\n",
    "    return haplotypes\n",
    "\n",
    "colors = [\"#781C86\", \"#571EA2\", \"#462EB9\", \"#3F47C9\", \"#3F63CF\", \"#447CCD\", \"#4C90C0\", \"#56A0AE\", \"#63AC9A\", \"#72B485\", \"#83BA70\", \"#96BD60\", \"#AABD52\", \"#BDBB48\", \"#CEB541\", \"#DCAB3C\", \"#E49938\", \"#E68133\", \"#E4632E\", \"#DF4327\", \"#DB2122\"]\n",
    "colors_lighter = [\"#A567AF\", \"#8F69C1\", \"#8474D1\", \"#7F85DB\", \"#7F97DF\", \"#82A8DD\", \"#88B5D5\", \"#8FC0C9\", \"#97C8BC\", \"#A1CDAD\", \"#ACD1A0\", \"#B9D395\", \"#C6D38C\", \"#D3D285\", \"#DECE81\", \"#E8C77D\", \"#EDBB7A\", \"#EEAB77\", \"#ED9773\", \"#EA816F\", \"#E76B6B\"]\n",
    "\n",
    "def stacked_trajectory_plot(xlabel=\"generation\"):\n",
    "    mpl.rcParams['font.size']=18\n",
    "    haplotypes = get_all_haplotypes()\n",
    "    trajectories = [get_trajectory(haplotype) for haplotype in haplotypes]\n",
    "    plt.stackplot(range(generations), trajectories, colors=colors_lighter, labels=haplotypes)\n",
    "    # plt.legend(bbox_to_anchor=(1.2, 1))\n",
    "    plt.ylim(0, 1)\n",
    "    plt.ylabel(\"frequency\")\n",
    "    plt.xlabel(xlabel)\n",
    "\n",
    "def get_snp_frequency(site, generation):\n",
    "    minor_allele_frequency = 0.0\n",
    "    pop_at_generation = history[generation]\n",
    "    for haplotype in pop_at_generation.keys():\n",
    "        allele = haplotype[site]\n",
    "        frequency = pop_at_generation[haplotype] / float(pop_size)\n",
    "        if allele != \"0\":\n",
    "            minor_allele_frequency += frequency\n",
    "    return minor_allele_frequency\n",
    "\n",
    "def get_snp_trajectory(site):\n",
    "    trajectory = [get_snp_frequency(site, gen) for gen in range(generations)]\n",
    "    return trajectory\n",
    "\n",
    "def get_all_snps():\n",
    "    snps = set()   \n",
    "    for generation in history:\n",
    "        for haplotype in generation:\n",
    "            for site in range(seq_length):\n",
    "                if haplotype[site] != \"0\":\n",
    "                    snps.add(site)\n",
    "    return snps\n",
    "\n",
    "def snp_trajectory_plot(xlabel=\"generation\"):\n",
    "    mpl.rcParams['font.size']=18\n",
    "    snps = get_all_snps()\n",
    "    trajectories = [get_snp_trajectory(snp) for snp in snps]\n",
    "    data = []\n",
    "    for trajectory, color in zip(trajectories, itertools.cycle(colors)):\n",
    "        data.append(range(generations))\n",
    "        data.append(trajectory)    \n",
    "        data.append(color)\n",
    "    fig = plt.plot(*data)   \n",
    "    plt.ylim(0, 1)\n",
    "    plt.ylabel(\"frequency\")\n",
    "    plt.xlabel(xlabel)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_drug(is_present):\n",
    "    # TODO: Maybe don't use globals here? We can pass the drug state into the step functions.\n",
    "    global fitness_base\n",
    "    global fitness_resistant\n",
    "    global fitness_fusion\n",
    "\n",
    "    if is_present:\n",
    "        fitness_base = fitness_drug['base']\n",
    "        fitness_resistant = fitness_drug['resistant']\n",
    "        fitness_fusion = fitness_drug['fusion']\n",
    "    else:\n",
    "        fitness_base = fitness_nodrug['base']\n",
    "        fitness_resistant = fitness_nodrug['resistant']\n",
    "        fitness_fusion = fitness_nodrug['fusion']\n",
    "\n",
    "    fitness[\"0\"] = fitness_base\n",
    "    fitness[\"1\"] = fitness_resistant\n",
    "    haplotypes = list(pop.keys())\n",
    "    for hap in haplotypes:\n",
    "        if len(hap) < 2: continue\n",
    "        fitness[hap] = get_fusion_fitness(hap)\n",
    "\n",
    "\n",
    "# Define functions for all the regimens for convenience\n",
    "\n",
    "drug_state = False\n",
    "\n",
    "def drug_off(generation):\n",
    "    set_drug(False)\n",
    "\n",
    "def drug_on(generation):\n",
    "    set_drug(True)\n",
    "\n",
    "def drug_on_off(generation):\n",
    "    if generation == 0:\n",
    "        set_drug(True)\n",
    "    elif generation == 250:\n",
    "        set_drug(False)\n",
    "\n",
    "\n",
    "def drug_off_on(generation):\n",
    "    if generation == 0:\n",
    "        set_drug(False)\n",
    "    elif generation == 250:\n",
    "        set_drug(True)\n",
    "\n",
    "def drug_off_on_off_on(generation):\n",
    "    global drug_state\n",
    "    if generation == 0:\n",
    "        drug_state = False\n",
    "    elif generation % 150 == 0:\n",
    "        drug_state = not drug_state\n",
    "        set_drug(drug_state)\n",
    "    set_drug(drug_state)\n",
    "\n",
    "regimens = {\n",
    "    0: drug_off,\n",
    "    1: drug_on,\n",
    "    2: drug_off_on,\n",
    "    3: drug_on_off,\n",
    "    4: drug_off_on_off_on\n",
    "}\n",
    "\n",
    "# Make it easy to save the regimen name as part of the output file.\n",
    "# TODO: Maybe just make these the keys for the regimen dict?\n",
    "regimen_names = {\n",
    "    0: 'drug_off',\n",
    "    1: 'drug_on',\n",
    "    2: 'drug_off_on',\n",
    "    3: 'drug_on_off',\n",
    "    4: 'drug_off_on_off_on'\n",
    "}\n",
    "\n",
    "def drug_step(generation, regimen_num):\n",
    "    regimens[regimen_num](generation)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "from pathlib import Path\n",
    "\n",
    "generations = 500\n",
    "trials = 10\n",
    "history = []\n",
    "\n",
    "def time_step():\n",
    "    mutation_step()\n",
    "    fusion_step()\n",
    "    offspring_step()\n",
    "\n",
    "def simulate(regimen):\n",
    "    clone_pop = dict(pop)\n",
    "    history.append(clone_pop)\n",
    "    for i in range(generations):\n",
    "        drug_step(i, regimen)\n",
    "        time_step()\n",
    "        clone_pop = dict(pop)\n",
    "        history.append(clone_pop)\n",
    "\n",
    "fractions = [0.9, 0.5, 0.1]\n",
    "for regimen_num in range(len(regimens)):\n",
    "    for fraction in fractions:\n",
    "        fraction_base = fraction\n",
    "\n",
    "        final_pops = []\n",
    "        fused_01_outcomes = []\n",
    "        fused_11_outcomes = []\n",
    "        for i in range(trials):\n",
    "            simulate(regimen_num)\n",
    "            \n",
    "            print('Final population: {}'.format(pop))\n",
    "            final_pops.append([val for pair in zip(pop.keys(), pop.values()) for val in pair])\n",
    "            \n",
    "            fused_01 = 0\n",
    "            fused_11 = 0\n",
    "            for (hap, count) in pop.items():\n",
    "                if hap == \"01\":\n",
    "                    fused_01 += count\n",
    "            fused_01_proportion = fused_01 / float(pop_size)\n",
    "            fused_01_outcomes.append(fused_01_proportion)\n",
    "\n",
    "        # Calculate average proportion over all trials\n",
    "        avg_01_proportion = 0.0\n",
    "        for proportion in fused_01_outcomes:\n",
    "            avg_01_proportion += proportion\n",
    "        avg_01_proportion = avg_01_proportion / len(fused_01_outcomes)\n",
    "        print('Mean 01 proportion after {} trials: {}'.format(trials, avg_01_proportion))\n",
    "\n",
    "        filename = \"r{}_output_{}.csv\".format(regimen_num, fraction_base)\n",
    "        filepath = Path('./output') / filename\n",
    "        with filepath.open('w') as csvfile:\n",
    "            csvwriter = csv.writer(csvfile)\n",
    "            csvwriter.writerow([fraction_base])\n",
    "            csvwriter.writerow([avg_01_proportion])\n",
    "            csvwriter.writerow([regimen_names[regimen_num]])\n",
    "            csvwriter.writerows(final_pops)\n",
    "\n",
    "# plt.figure(num=None, figsize=(14, 14), dpi=80, facecolor='w', edgecolor='k')\n",
    "# plt.subplot2grid((2,2), (0,0), colspan=2)\n",
    "# stacked_trajectory_plot()\n",
    "# #plt.subplot2grid((3,2), (1,0), colspan=2)\n",
    "# #snp_trajectory_plot()\n",
    "# plt.subplot2grid((2,2), (1,0))\n",
    "# diversity_plot()\n",
    "# plt.subplot2grid((2,2), (1,1))\n",
    "# divergence_plot()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plotting\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "\n",
    "plt.figure(figsize=(10, 7), dpi=80)\n",
    "gs = gridspec.GridSpec(3, 5)\n",
    "gs.update(wspace=0, hspace=0) # set the spacing between axes. \n",
    "\n",
    "\n",
    "results = [file for file in Path('./output').iterdir() if file.is_file() and file.name.startswith('r')]\n",
    "results.sort()\n",
    "for count, file in enumerate(results):\n",
    "   row_num = count % 3\n",
    "   col_num = int(count / 3)\n",
    "\n",
    "   with file.open('r') as csvfile:\n",
    "      reader = csv.reader(csvfile)\n",
    "      rows = [row for row in reader]\n",
    "      starting_fraction = float(rows[0][0])\n",
    "      proportions = [float(val) for val in rows[1]]\n",
    "      proportion_fused = sum(proportions)\n",
    "\n",
    "      regimen_name = rows[2][0]\n",
    "      print('Plotting ', regimen_name)\n",
    "\n",
    "      ax = plt.subplot(gs[row_num, col_num])\n",
    "      ax.set_facecolor((0.3, 0.3, 0.3, proportion_fused / 0.02))\n",
    "      ax.text(0.5, 0.5, '01: {:.5f}'.format(proportions[0]),\n",
    "         bbox={'facecolor':'white','alpha':1,'edgecolor':'none','pad':1},\n",
    "         ha='center', va='center')\n",
    "\n",
    "      # Add the drug regimen to the bottom row label.\n",
    "      if row_num == 2 or col_num == 0:\n",
    "         if row_num == 2:\n",
    "            ax.set_xlabel(regimen_name)\n",
    "            ax.xaxis.set_ticklabels([])\n",
    "            ax.tick_params(axis='both', which='both', bottom=False, left=False, labelleft=False)\n",
    "         if col_num == 0:\n",
    "            # Add the proportion to the first column label\n",
    "            ax.set_ylabel('Base: {}'.format(starting_fraction))\n",
    "            ax.yaxis.set_ticklabels([])\n",
    "            ax.xaxis.set_ticklabels([])\n",
    "            ax.tick_params(axis='both', which='both', bottom=False, left=False)\n",
    "      else:\n",
    "         ax.tick_params(axis='both', which='both', bottom=False, labelbottom=False, left=False, labelleft=False)\n",
    "\n",
    "   row_num += 1\n",
    "   col_num += 1\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
  },
  "kernelspec": {
   "display_name": "Python 3.9.10 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
